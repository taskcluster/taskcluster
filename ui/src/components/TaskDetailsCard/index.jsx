import React, { Component, Fragment } from 'react';
import { arrayOf, shape, string } from 'prop-types';
import { CopyToClipboard } from 'react-copy-to-clipboard';
import deepSortObject from 'deep-sort-object';
import Code from '@mozilla-frontend-infra/components/Code';
import Label from '@mozilla-frontend-infra/components/Label';
import { withStyles } from '@material-ui/core/styles';
import Card from '@material-ui/core/Card';
import CardContent from '@material-ui/core/CardContent';
import Typography from '@material-ui/core/Typography';
import Collapse from '@material-ui/core/Collapse';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import ChevronUpIcon from 'mdi-react/ChevronUpIcon';
import ChevronDownIcon from 'mdi-react/ChevronDownIcon';
import ContentCopyIcon from 'mdi-react/ContentCopyIcon';
import LinkIcon from 'mdi-react/LinkIcon';
import OpenInNewIcon from 'mdi-react/OpenInNewIcon';
import DateDistance from '../DateDistance';
import StatusLabel from '../StatusLabel';
import { task } from '../../utils/prop-types';
import urls from '../../utils/urls';
import Link from '../../utils/Link';

@withStyles(theme => ({
  headline: {
    paddingLeft: theme.spacing.triple,
    paddingRight: theme.spacing.triple,
  },
  cardContent: {
    paddingLeft: 0,
    paddingRight: 0,
    paddingTop: theme.spacing.double,
    paddingBottom: theme.spacing.double,
    '&:last-child': {
      paddingBottom: theme.spacing.double,
    },
  },
  sourceHeadline: {
    textOverflow: 'ellipsis',
    overflowX: 'hidden',
    whiteSpace: 'nowrap',
  },
  sourceHeadlineText: {
    flex: 1,
  },
  listItemButton: {
    ...theme.mixins.listItemButton,
  },
  pre: {
    margin: 0,
    fontSize: theme.typography.body2.fontSize,
  },
  unorderedList: {
    ...theme.mixins.unorderedList,
  },
}))
/**
 * Render information in a card layout about a task.
 */
export default class TaskDetailsCard extends Component {
  static defaultProps = {
    dependentTasks: null,
  };

  static propTypes = {
    /**
     * A GraphQL task response.
     */
    task: task.isRequired,
    /**
     * A collection of GraphQL task responses associated with the given task.
     */
    dependentTasks: arrayOf(
      shape({
        taskId: string,
        status: shape({
          state: string,
        }),
        metadata: shape({
          name: string,
        }),
      })
    ),
  };

  state = {
    showPayload: false,
    showMore: false,
  };

  handleTogglePayload = () => {
    this.setState({ showPayload: !this.state.showPayload });
  };

  handleToggleMore = () => {
    this.setState({ showMore: !this.state.showMore });
  };

  render() {
    const { classes, task, dependentTasks } = this.props;
    const { showPayload, showMore } = this.state;
    const isExternal = task.metadata.source.startsWith('https://');
    const payload = deepSortObject(task.payload);

    return (
      <Card raised>
        <div>
          <CardContent classes={{ root: classes.cardContent }}>
            <Typography variant="h5" className={classes.headline}>
              Task Details
            </Typography>
            <List>
              <ListItem>
                <ListItemText
                  primary="State"
                  secondary={<StatusLabel state={task.status.state} />}
                />
              </ListItem>
              <CopyToClipboard
                title={`${task.created} (Copy)`}
                text={task.created}>
                <ListItem button className={classes.listItemButton}>
                  <ListItemText
                    primary="Created"
                    secondary={<DateDistance from={task.created} />}
                  />
                  <ContentCopyIcon />
                </ListItem>
              </CopyToClipboard>
              <ListItem>
                <ListItemText
                  primary="Provisioner"
                  secondary={task.provisionerId}
                />
              </ListItem>
              <ListItem
                title="View Workers"
                button
                className={classes.listItemButton}
                component={Link}
                to={`/provisioners/${task.provisionerId}/worker-types/${task.workerType}`}>
                <ListItemText
                  primary="Worker Type"
                  secondary={task.workerType}
                />
                <LinkIcon />
              </ListItem>
              <ListItem
                button
                className={classes.listItemButton}
                onClick={this.handleTogglePayload}>
                <ListItemText primary="Payload" />
                {showPayload ? <ChevronUpIcon /> : <ChevronDownIcon />}
              </ListItem>
              <Collapse in={showPayload} timeout="auto">
                <List component="div" disablePadding>
                  <ListItem>
                    <ListItemText
                      disableTypography
                      primary={
                        <Code language="json">
                          {JSON.stringify(payload, null, 2)}
                        </Code>
                      }
                    />
                  </ListItem>
                </List>
              </Collapse>
              <ListItem
                button
                className={classes.listItemButton}
                component="a"
                href={urls.api('queue', 'v1', `task/${task.taskId}`)}
                target="_blank"
                rel="noopener noreferrer">
                <ListItemText primary="View task definition" />
                <OpenInNewIcon />
              </ListItem>
              <ListItem
                button
                className={classes.listItemButton}
                onClick={this.handleToggleMore}>
                <ListItemText primary={showMore ? 'Less...' : 'More...'} />
                {showMore ? <ChevronUpIcon /> : <ChevronDownIcon />}
              </ListItem>
            </List>
            <Collapse in={showMore} timeout="auto">
              <List>
                <ListItem
                  button
                  className={classes.listItemButton}
                  component="a"
                  href={task.metadata.source}
                  target="_blank"
                  rel="noopener noreferrer">
                  <ListItemText
                    className={classes.sourceHeadlineText}
                    classes={{ secondary: classes.sourceHeadline }}
                    primary="Source"
                    secondary={task.metadata.source}
                    title={task.metadata.source}
                  />
                  {isExternal ? <OpenInNewIcon /> : <LinkIcon />}
                </ListItem>
                <ListItem>
                  <ListItemText
                    primary="Retries Left"
                    secondary={`${task.status.retriesLeft} of ${task.retries}`}
                  />
                </ListItem>
                <CopyToClipboard
                  title={`${task.deadline} (Copy)`}
                  text={task.deadline}>
                  <ListItem button className={classes.listItemButton}>
                    <ListItemText
                      primary="Deadline"
                      secondary={
                        <DateDistance
                          from={task.deadline}
                          offset={task.created}
                        />
                      }
                    />
                    <ContentCopyIcon />
                  </ListItem>
                </CopyToClipboard>
                <CopyToClipboard
                  title={`${task.expires} (Copy)`}
                  text={task.expires}>
                  <ListItem button className={classes.listItemButton}>
                    <ListItemText
                      primary="Expires"
                      secondary={<DateDistance from={task.expires} />}
                    />
                    <ContentCopyIcon />
                  </ListItem>
                </CopyToClipboard>
                <ListItem>
                  <ListItemText
                    primary="Priority"
                    secondary={
                      <Label mini status="info">
                        {task.priority}
                      </Label>
                    }
                  />
                </ListItem>
                {dependentTasks && dependentTasks.length ? (
                  <Fragment>
                    <ListItem>
                      <ListItemText
                        primary="Dependencies"
                        secondary={
                          <Fragment>
                            This task will be scheduled when
                            <strong>
                              <em> dependencies </em>
                            </strong>
                            are
                            {task.requires === 'ALL_COMPLETED' ? (
                              <Fragment>
                                &nbsp;
                                <code>all-completed</code> successfully.
                              </Fragment>
                            ) : (
                              <Fragment>
                                &nbsp;
                                <code>all-resolved</code> with any resolution.
                              </Fragment>
                            )}
                          </Fragment>
                        }
                      />
                    </ListItem>
                    <List dense disablePadding>
                      {dependentTasks.map(task => (
                        <ListItem
                          button
                          component={Link}
                          className={classes.listItemButton}
                          to={`/tasks/${task.taskId}`}
                          key={task.taskId}
                          title="View Task">
                          <StatusLabel state={task.status.state} />
                          <ListItemText primary={task.metadata.name} />
                          <LinkIcon />
                        </ListItem>
                      ))}
                    </List>
                  </Fragment>
                ) : (
                  <ListItem>
                    <ListItemText
                      primary="Dependencies"
                      secondary={<em>n/a</em>}
                    />
                  </ListItem>
                )}
                <ListItem>
                  <ListItemText
                    primary="Scheduler ID"
                    secondary={
                      task.schedulerId === '-' ? <em>n/a</em> : task.schedulerId
                    }
                  />
                </ListItem>
                <ListItem>
                  <ListItemText
                    disableTypography
                    primary={
                      <Typography variant="subtitle1">Scopes</Typography>
                    }
                    secondary={
                      task.scopes.length ? (
                        <ul className={classes.unorderedList}>
                          {task.scopes.map(scope => (
                            <li key={scope}>
                              <Typography
                                component="span"
                                color="textSecondary">
                                {scope}
                              </Typography>
                            </li>
                          ))}
                        </ul>
                      ) : (
                        <em>n/a</em>
                      )
                    }
                  />
                </ListItem>
                <ListItem>
                  <ListItemText
                    disableTypography
                    primary={
                      <Typography variant="subtitle1">Routes</Typography>
                    }
                    secondary={
                      task.routes.length ? (
                        <ul className={classes.unorderedList}>
                          {task.routes.map(route => (
                            <li key={route}>
                              <Typography
                                component="span"
                                color="textSecondary">
                                {route}
                              </Typography>
                            </li>
                          ))}
                        </ul>
                      ) : (
                        <em>n/a</em>
                      )
                    }
                  />
                </ListItem>
                {Object.keys(task.extra).length !== 0 && (
                  <ListItem>
                    <ListItemText
                      disableTypography
                      primary={
                        <Typography variant="subtitle1">Extra</Typography>
                      }
                      secondary={
                        <Code language="json">
                          {JSON.stringify(task.extra, null, 2)}
                        </Code>
                      }
                    />
                  </ListItem>
                )}
              </List>
            </Collapse>
          </CardContent>
        </div>
      </Card>
    );
  }
}
