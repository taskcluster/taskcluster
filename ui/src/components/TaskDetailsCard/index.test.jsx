import React from 'react';
import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import TaskDetailsCard from './index';

it('should render TaskDetailsCard', () => {
  const { asFragment } = render(
    <MemoryRouter keyLength={0}>
      <TaskDetailsCard
        task={{
          taskId: 'taskId',
          status: {
            taskId: 'taskId',
            provisioinerId: 'p1',
            workerType: 'workerType',
            taskGroupId: 'tg1',
            state: 'RUNNING',
          },
          taskQueueId: 'task/queue/id',
          created: '2022-02-15T12:00:00.000Z',
          deadline: '2022-05-15T12:00:00.000Z',
          expires: '2023-02-15T12:00:00.000Z',
          scopes: ['scopes'],
          routes: [],
          payload: {
            command: [
              '/bin/bash',
              '-c',
              'for ((i=1;i<=600;i++)); do echo $i; sleep 1; done',
            ],
            image: 'ubuntu:latest',
            maxRunTime: 630,
          },
          metadata: {
            source: 'https://test.taskcluster-dev.net/task/taskId',
          },
          extra: {},
        }}
        dependents={[]}
        onDependentsPageChange={jest.fn()}
        objectContent={{}}
      />
    </MemoryRouter>
  );

  expect(asFragment()).toMatchSnapshot();
});
