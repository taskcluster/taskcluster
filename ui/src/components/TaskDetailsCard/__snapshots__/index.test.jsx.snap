// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render TaskDetailsCard 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiCard-root MuiPaper-elevation8 MuiPaper-rounded"
  >
    <div>
      <div
        class="MuiCardContent-root TaskDetailsCard-cardContent-2"
      >
        <h5
          class="MuiTypography-root TaskDetailsCard-headline-1 MuiTypography-h5"
        >
          Task Details
        </h5>
        <ul
          class="MuiList-root MuiList-padding"
        >
          <li
            class="MuiListItem-root MuiListItem-gutters"
          >
            <div
              class="MuiListItemText-root MuiListItemText-multiline"
            >
              <span
                class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
              >
                State
              </span>
              <p
                class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
              >
                <button
                  class="MuiButtonBase-root MuiButton-root MuiButton-text Label-mini-17 MuiButton-textSizeSmall MuiButton-sizeSmall Label-dense-16 Mui-disabled Label-info-23 Label-disabled-18 Mui-disabled"
                  disabled=""
                  tabindex="-1"
                  type="button"
                >
                  <span
                    class="MuiButton-label"
                  >
                    RUNNING
                  </span>
                </button>
              </p>
            </div>
          </li>
          <div
            aria-disabled="false"
            class="MuiButtonBase-root MuiListItem-root makeStyles-listItemButtonRoot-24 MuiListItem-gutters MuiListItem-button"
            role="button"
            tabindex="0"
            title="taskId (Copy)"
          >
            <div
              class="MuiListItemText-root MuiListItemText-multiline"
            >
              <span
                class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
              >
                Task ID
              </span>
              <p
                class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
              >
                taskId
              </p>
            </div>
            <svg
              class="mdi-icon "
              fill="currentColor"
              height="24"
              viewBox="0 0 24 24"
              width="24"
            >
              <path
                d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root"
            />
          </div>
          <div
            aria-disabled="false"
            class="MuiButtonBase-root MuiListItem-root makeStyles-listItemButtonRoot-24 MuiListItem-gutters MuiListItem-button"
            role="button"
            tabindex="0"
            title="2022-02-15T12:00:00.000Z (Copy)"
          >
            <div
              class="MuiListItemText-root MuiListItemText-multiline"
            >
              <span
                class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
              >
                Created
              </span>
              <p
                class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
              >
                2 days ago
              </p>
            </div>
            <svg
              class="mdi-icon "
              fill="currentColor"
              height="24"
              viewBox="0 0 24 24"
              width="24"
            >
              <path
                d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root"
            />
          </div>
          <a
            href="/provisioners/task/worker-types/queue"
          >
            <div
              aria-disabled="false"
              class="MuiButtonBase-root MuiListItem-root TaskDetailsCard-listItemButton-5 MuiListItem-gutters MuiListItem-button"
              role="button"
              tabindex="0"
              title="View Workers"
            >
              <div
                class="MuiListItemText-root MuiListItemText-multiline"
              >
                <span
                  class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                >
                  Task Queue ID
                </span>
                <p
                  class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
                >
                  task/queue/id
                </p>
              </div>
              <svg
                class="mdi-icon "
                fill="currentColor"
                height="24"
                viewBox="0 0 24 24"
                width="24"
              >
                <path
                  d="M3.9,12C3.9,10.29 5.29,8.9 7,8.9H11V7H7A5,5 0 0,0 2,12A5,5 0 0,0 7,17H11V15.1H7C5.29,15.1 3.9,13.71 3.9,12M8,13H16V11H8V13M17,7H13V8.9H17C18.71,8.9 20.1,10.29 20.1,12C20.1,13.71 18.71,15.1 17,15.1H13V17H17A5,5 0 0,0 22,12A5,5 0 0,0 17,7Z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root"
              />
            </div>
          </a>
          <div
            aria-disabled="false"
            class="MuiButtonBase-root MuiListItem-root TaskDetailsCard-listItemButton-5 MuiListItem-gutters MuiListItem-button"
            role="button"
            tabindex="0"
          >
            <div
              class="MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
              >
                Payload
              </span>
            </div>
            <svg
              class="mdi-icon "
              fill="currentColor"
              height="24"
              viewBox="0 0 24 24"
              width="24"
            >
              <path
                d="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root"
            />
          </div>
          <div
            class="MuiCollapse-root MuiCollapse-hidden"
            style="min-height: 0px;"
          >
            <div
              class="MuiCollapse-wrapper"
            >
              <div
                class="MuiCollapse-wrapperInner"
              >
                <div
                  class="MuiList-root"
                >
                  <li
                    class="MuiListItem-root MuiListItem-gutters"
                  >
                    <div
                      class="MuiListItemText-root"
                    >
                      <div
                        class="MuiTypography-root MuiTypography-body1"
                      >
                        <label
                          class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 MuiGrid-align-items-xs-center"
                        >
                          <div
                            class="MuiGrid-root MuiGrid-item"
                          >
                            JSON
                          </div>
                          <div
                            class="MuiGrid-root MuiGrid-item"
                          >
                            <span
                              class="MuiSwitch-root"
                            >
                              <span
                                aria-disabled="false"
                                class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-25 MuiSwitch-switchBase MuiSwitch-colorSecondary"
                              >
                                <span
                                  class="MuiIconButton-label"
                                >
                                  <input
                                    class="PrivateSwitchBase-input-28 MuiSwitch-input"
                                    name="yaml_json_switch"
                                    type="checkbox"
                                    value=""
                                  />
                                  <span
                                    class="MuiSwitch-thumb"
                                  />
                                </span>
                                <span
                                  class="MuiTouchRipple-root"
                                />
                              </span>
                              <span
                                class="MuiSwitch-track"
                              />
                            </span>
                          </div>
                          <div
                            class="MuiGrid-root MuiGrid-item"
                          >
                            YAML
                          </div>
                        </label>
                      </div>
                      <pre
                        class="language-json TaskDetailsCard-payload-9"
                      >
                        <code>
                          <span
                            class="hljs-punctuation"
                          >
                            {
                          </span>
                          
  
                          <span
                            class="hljs-attr"
                          >
                            "command"
                          </span>
                          <span
                            class="hljs-punctuation"
                          >
                            :
                          </span>
                           
                          <span
                            class="hljs-punctuation"
                          >
                            [
                          </span>
                          
    
                          <span
                            class="hljs-string"
                          >
                            "/bin/bash"
                          </span>
                          <span
                            class="hljs-punctuation"
                          >
                            ,
                          </span>
                          
    
                          <span
                            class="hljs-string"
                          >
                            "-c"
                          </span>
                          <span
                            class="hljs-punctuation"
                          >
                            ,
                          </span>
                          
    
                          <span
                            class="hljs-string"
                          >
                            "for ((i=1;i&lt;=600;i++)); do echo $i; sleep 1; done"
                          </span>
                          
  
                          <span
                            class="hljs-punctuation"
                          >
                            ]
                          </span>
                          <span
                            class="hljs-punctuation"
                          >
                            ,
                          </span>
                          
  
                          <span
                            class="hljs-attr"
                          >
                            "image"
                          </span>
                          <span
                            class="hljs-punctuation"
                          >
                            :
                          </span>
                           
                          <span
                            class="hljs-string"
                          >
                            "ubuntu:latest"
                          </span>
                          <span
                            class="hljs-punctuation"
                          >
                            ,
                          </span>
                          
  
                          <span
                            class="hljs-attr"
                          >
                            "maxRunTime"
                          </span>
                          <span
                            class="hljs-punctuation"
                          >
                            :
                          </span>
                           
                          <span
                            class="hljs-number"
                          >
                            630
                          </span>
                          

                          <span
                            class="hljs-punctuation"
                          >
                            }
                          </span>
                        </code>
                      </pre>
                    </div>
                  </li>
                </div>
              </div>
            </div>
          </div>
          <a
            class="TaskDetailsCard-listItemButton-5"
            href="/tasks/taskId/definition"
          >
            <div
              aria-disabled="false"
              class="MuiButtonBase-root MuiListItem-root TaskDetailsCard-listItemButton-5 MuiListItem-gutters MuiListItem-button"
              role="button"
              tabindex="0"
            >
              <div
                class="MuiListItemText-root"
              >
                <span
                  class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                >
                  Full Task Definition
                </span>
              </div>
              <span
                class="MuiTouchRipple-root"
              />
            </div>
          </a>
          <a
            aria-disabled="false"
            class="MuiButtonBase-root MuiListItem-root TaskDetailsCard-listItemButton-5 MuiListItem-gutters MuiListItem-button"
            href="https://test.taskcluster-dev.net/task/taskId"
            rel="noopener noreferrer"
            tabindex="0"
            target="_blank"
          >
            <div
              class="MuiListItemText-root TaskDetailsCard-sourceHeadlineText-4 MuiListItemText-multiline"
              title="https://test.taskcluster-dev.net/task/taskId"
            >
              <span
                class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
              >
                Source
              </span>
              <p
                class="MuiTypography-root MuiListItemText-secondary TaskDetailsCard-sourceHeadline-3 MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
              >
                https://test.taskcluster-dev.net/task/taskId
              </p>
            </div>
            <svg
              class="mdi-icon "
              fill="currentColor"
              height="24"
              viewBox="0 0 24 24"
              width="24"
            >
              <path
                d="M14,3V5H17.59L7.76,14.83L9.17,16.24L19,6.41V10H21V3M19,19H5V5H12V3H5C3.89,3 3,3.9 3,5V19A2,2 0 0,0 5,21H19A2,2 0 0,0 21,19V12H19V19Z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root"
            />
          </a>
          <li
            class="MuiListItem-root MuiListItem-gutters"
          >
            <div
              class="MuiListItemText-root MuiListItemText-multiline"
            >
              <span
                class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
              >
                Retries Left
              </span>
              <p
                class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
              >
                undefined of undefined
              </p>
            </div>
          </li>
          <div
            aria-disabled="false"
            class="MuiButtonBase-root MuiListItem-root makeStyles-listItemButtonRoot-24 MuiListItem-gutters MuiListItem-button"
            role="button"
            tabindex="0"
            title="2022-05-15T12:00:00.000Z (Copy)"
          >
            <div
              class="MuiListItemText-root MuiListItemText-multiline"
            >
              <span
                class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
              >
                Deadline
              </span>
              <p
                class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
              >
                in 3 months (3 months later)
              </p>
            </div>
            <svg
              class="mdi-icon "
              fill="currentColor"
              height="24"
              viewBox="0 0 24 24"
              width="24"
            >
              <path
                d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root"
            />
          </div>
          <div
            aria-disabled="false"
            class="MuiButtonBase-root MuiListItem-root makeStyles-listItemButtonRoot-24 MuiListItem-gutters MuiListItem-button"
            role="button"
            tabindex="0"
            title="2023-02-15T12:00:00.000Z (Copy)"
          >
            <div
              class="MuiListItemText-root MuiListItemText-multiline"
            >
              <span
                class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
              >
                Expires
              </span>
              <p
                class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
              >
                in 1 year
              </p>
            </div>
            <svg
              class="mdi-icon "
              fill="currentColor"
              height="24"
              viewBox="0 0 24 24"
              width="24"
            >
              <path
                d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root"
            />
          </div>
          <li
            class="MuiListItem-root MuiListItem-gutters"
          >
            <div
              class="MuiListItemText-root MuiListItemText-multiline"
            >
              <span
                class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
              >
                Priority
              </span>
              <p
                class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
              >
                <button
                  class="MuiButtonBase-root MuiButton-root MuiButton-text Label-mini-17 MuiButton-textSizeSmall MuiButton-sizeSmall Label-dense-16 Mui-disabled Label-info-23 Label-disabled-18 Mui-disabled"
                  disabled=""
                  tabindex="-1"
                  type="button"
                >
                  <span
                    class="MuiButton-label"
                  />
                </button>
              </p>
            </div>
          </li>
          <li
            class="MuiListItem-root MuiListItem-gutters"
          >
            <div
              class="MuiListItemText-root MuiListItemText-multiline"
            >
              <span
                class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
              >
                Dependencies
              </span>
              <p
                class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
              >
                <em>
                  n/a
                </em>
              </p>
            </div>
          </li>
          <li
            class="MuiListItem-root MuiListItem-gutters"
          >
            <div
              class="MuiListItemText-root MuiListItemText-multiline"
            >
              <span
                class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
              >
                Dependents
              </span>
              <p
                class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
              >
                <em>
                  n/a
                </em>
              </p>
            </div>
          </li>
          <li
            class="MuiListItem-root MuiListItem-gutters"
          >
            <div
              class="MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
              >
                Project ID
              </span>
            </div>
          </li>
          <li
            class="MuiListItem-root MuiListItem-gutters"
          >
            <div
              class="MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
              >
                Scheduler ID
              </span>
            </div>
          </li>
          <li
            class="MuiListItem-root MuiListItem-gutters"
          >
            <div
              class="MuiListItemText-root MuiListItemText-multiline"
            >
              <h6
                class="MuiTypography-root MuiTypography-subtitle1"
              >
                Scopes
              </h6>
              <ul
                class="TaskDetailsCard-unorderedList-7"
              >
                <li>
                  <span
                    class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
                  >
                    scopes
                  </span>
                </li>
              </ul>
            </div>
          </li>
          <li
            class="MuiListItem-root MuiListItem-gutters"
          >
            <div
              class="MuiListItemText-root MuiListItemText-multiline"
            >
              <h6
                class="MuiTypography-root MuiTypography-subtitle1"
              >
                Routes
              </h6>
              <em>
                n/a
              </em>
            </div>
          </li>
        </ul>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
